app_allocator/classes/ordered_queues.py

```diff
@@ -0,0 +1,161 @@
+                for feature in OrderedQueues.features]
+
+    def work_left(self):
+        for queue in self.queues:
+            if queue.work_left() > 0:
+                return True
+        return False
+
+    def process_judge_events(self, events):
+        for event in events:
+            action = event.fields.get("action")
+            judge = event.fields.get("subject")
+            application = event.fields.get("object")
+            if action and judge and application:
+                self._update_needs(action, judge, application)
+
+    def _update_needs(self, action, judge, application):
+        if action in OrderedQueues.relevant_actions:
+            application.process_judge_action(action, judge)
+            needs = self.application_needs[application]
+            if needs:
+                new_needs = _calc_new_needs(needs, action, judge)
+                self.application_needs[application] = new_needs
+                self._remove_assignment(application, judge)
+                self._queue_for_needs(application, at_front=(action == "pass"))
+
+    def _remove_assignment(self, application, judge):
+        queue = self.application_queues[application]
+        queue.remove_assignment(application, judge)
+
+    def find_one_application(self, judge):
+        queue, value = self._find_best_queue(judge)
+        if queue:
+            application = self._next_item(queue, judge)
+            queue.move_to_end(application)
+            Event(action="select", subject=judge, object=application,
+                  description="{queue}: {value}".format(queue=str(queue),
+                                                        value=value))
+            return application
+        return None
+
+    def _find_best_queue(self, judge):
+        best_queues = []
+        best_value = -1
+        for queue in self.queues:
+            best_value, best_queues = _evaluate_queue_for_judge(queue,
+                                                                judge,
+                                                                best_value,
+                                                                best_queues)
+        if best_queues:
+            return choice(best_queues), best_value
+        return None, 0
+
+    def _next_item(self, queue, judge):
+        application = queue.assign_next_application(judge)
+        return application
+
+    def assess(self):
+        for queue in self.queues:
+            remaining = queue.remaining()
+            if remaining > 0:
+                description = "{remaining} item(s) left: {example}".format(
+                    remaining=remaining,
+                    example=queue.items[0].properties)
+                Event(action="fail",
+                      subject=queue,
+                      description=description)
+            else:
+                Event(action="complete",
+                      subject=queue)
+#         self.assess_zscore()
```

Comment:
commented code - can be removed?

Reply:
This is Z-score debugging code that I'm not quite ready to let go.